yourgov/lib/useSessionFormState.ts

```diff
@@ -9,38 +12,126 @@ type SetStateValue<TValue> = TValue | ((prevState: TValue) => TValue);
 export const safeSessionStorage =
 	'sessionStorage' in globalThis ? globalThis.sessionStorage : undefined;
 
-export function useSessionFormState<FormState>(
+/**
+ * This hook sets, stores, and synchronises all state between the session
+ * storage and the global React state.
+ *
+ * TODO: single responsibility
+ */
+export function useSessionFormState<GlobalState extends DeepPartial<FormState>>(
 	key: string,
-	initialValue: Partial<FormState>
+	initialValue: Partial<GlobalState>
 ) {
 	const [hasSynced, setHasSynced] = useState(false);
-	const [localFormState, setLocalFormState] = useState(initialValue);
+	const [globalState, setGlobalState] = useState(initialValue);
 
 	useEffect(() => {
 		if (!safeSessionStorage) return;
 		const value = safeSessionStorage.getItem(key);
-		const parsedValue = value ? (JSON.parse(value) as FormState) : null;
-		if (parsedValue) setLocalFormState(parsedValue);
+		const parsedValue = value ? (JSON.parse(value) as GlobalState) : null;
+		if (parsedValue) setGlobalState(parsedValue);
 		setHasSynced(true);
 	}, [initialValue, key]);
 
-	const localFormStateRef = useRef(localFormState);
-	localFormStateRef.current = localFormState;
+	const globalStateRef = useRef(globalState);
+	globalStateRef.current = globalState;
 
-	const setState = useCallback(
-		(value: SetStateValue<Partial<FormState>>) => {
-			function writeStorage(key: string, value: Partial<FormState>) {
+	// TODO: migrate this to a function called `setGlobalState` that calls a function `syncGlobalStateAndSessionStorage` rather than doing everything in the 'session storage manager' hook
+	const setAndSyncGlobalStateAndSessionStorage = useCallback(
+		(value: SetStateValue<Partial<GlobalState>>) => {
+			function writeStorage(key: string, value: Partial<GlobalState>) {
 				safeSessionStorage?.setItem(key, JSON.stringify(value));
-				setLocalFormState(value);
+				setGlobalState(value);
 			}
 			value instanceof Function
-				? writeStorage(key, value(localFormStateRef.current))
+				? writeStorage(key, value(globalStateRef.current))
 				: writeStorage(key, value);
 		},
 		[key]
 	);
 
-	return [hasSynced, localFormState, setState] as const;
+	const stateSettersPerStep = useMemo(
+		() => ({
+			step1SetState: (newState: Partial<TaskFormState['step1']>) => {
+				setAndSyncGlobalStateAndSessionStorage((prevState) => {
+					return {
+						...prevState,
+						task: {
+							...prevState.task,
+							step1: {
+								...prevState.task?.step1, // User details, e.g. name and email are saved against step 1 but not always included during form submisssions for step 1 - TODO: either migrate this data to hidden form fields in step 1 or include it separately in the confirm and submit page
```

Comment:
Interesting, this is referencing server form actions?

Reply:
I'm not sure I fully understand what you mean here but the user details are saved against `task.step1`, presumably because you can change them in a step 1 sub-step. This isn't how apps work WRT user details, though. If the data was different, e.g. some flow-specific contact details and we wanted to demonstrate a pattern for sub-step state management, then we'd probably want hidden fields (which would need to be mirrored in the sub-step) or a nested object in state to reflect the nested sub-step. So, I think this just needs more thought but will do for now 🤷